### DISCUSSION TITLE
Hotkeys for Specific Patterns

### DISCUSSION THREAD

**Discussion by FelipeRearden** [jglev/obsidian-apply-patterns-plugin#5] [Ideas]
[More info about me here](https://github.com/publicus/obsidian-apply-patterns-plugin/discussions/3)

Hello @publicus !!!

My workflow in Obsidian is 99% with keyboard and hotkeys. 
In my opinion, would be amazing to be able to create a specific hotkey to each one of the Patterns in Command Palette

To make this possible, the pattern option prompt `window B` should be appear `after` the user select the pattern option `window A`.

**From the example in your git:**

`window B` Command Palette —> 
- Apply Pattern: “Set Open task as Complete”
- Apply Pattern: “Reschedule closed task”
![Screen Shot 2021-08-10 at 08 48 35](https://user-images.githubusercontent.com/85180696/128862171-a27395e8-183a-4026-9916-88c27c551f44.png)

`window A` User selected —> Apply Pattern: “Set Open task as Complete”
- Apply Pattern to the whole line
- Apply Pattern to selection 
![Screen Shot 2021-08-10 at 08 48 02](https://user-images.githubusercontent.com/85180696/128861965-06860f9a-afec-421b-b43d-509c03eb61a7.png)

`window B` should appear before `window A` in my opinion. This way I can apply the hotkey to the specific pattern them decide about the range (`whole line` or `to selection`)

Let me know if you need more information.

Thanks and have a great day!

**Comment by jglev** [comment]
Thank you for writing, @FelipeRearden!

This is something that I'd been thinking about, as well. I hadn't targeted it yet because to my knowledge, there's no way to un-register a command in Obsidian after registering it, short of reloading Obsidian. Thus, for commands with names that the user might change at any given time, I wanted to think more about how to apply this.

I've come up with what I think is a flexible way that would allow exposing individual Patterns as commands, as well as grouping Patterns together into new command menus. The draft is in #11; are you familiar with how to test that? If so, would you be willing to give it a try? (If not, no worries, of course!)

In the Settings tab, I've added a new "Commands" section. Each command has a name, and a search string for Pattern names.

![image](https://user-images.githubusercontent.com/3667562/129076111-ee3e883f-e2ee-428a-acd1-d8a67fd9b1a2.png)

If a command matches just one Pattern, it will run that Pattern. If it matches multiple Patterns, it will pop up a searchable menu asking which Pattern to run. Each command can be set to run on whole lines, the whole document, and/or just the selection. (So the setting in the screenshot, which is set to both Lines and Document, would create two new commands in Obsidian: `Apply Patterns: Task Management on whole lines`, and `Apply Patterns: Task Management on whole document`, each of which could have a hotkey set for it).

This allows setting up hotkeys for individual Patterns for different scopes (tying into your suggestion in #8), as well as for collections of Patterns that share a common theme or context. Does that make good sense, as I'm putting it?

**Reply by FelipeRearden** [reply]
Hello again and again @publicus !!!


Now I understand the reason. Thanks for explain to me :)

Woowww this is an amazing workaround, it's brilliant :)

I have nothing to say only thank you for making this change, it's absolutely amazing how you create an new feature to specific hotkeys, patterns options. Every workflow is protected now !!!!!

When 1.2.0 appears in my Obsidian I'm gonna take a look and provide feedback. I'm so excited to take a look at this :) 

Thank you very much!!!!

**Reply by jglev** [reply]
Thank you for your words! :blush: 

I've released 1.2.0 just now, and appreciate you taking a look!

**Reply by FelipeRearden** [reply]
> Thank you for your words! 😊
> 
> I've released 1.2.0 just now, and appreciate you taking a look!

Just to let you now that is not avaliable for me yet. I message you when I get the new version  :)
![Screen Shot 2021-08-12 at 11 39 12](https://user-images.githubusercontent.com/85180696/129216598-7d46a52c-1b8a-4d77-9b16-8cc6653fcff9.png)

**Reply by jglev** [reply]
@FelipeRearden, thank you for letting me know! I'd forgotten to update the version number; oops. I've re-released it just now; is it appearing for you now?

**Reply by FelipeRearden** [reply]
@publicus 

yes! It is available now :)
![C39932DB-A631-48F6-A11A-5C2D8451E943](https://user-images.githubusercontent.com/85180696/129225118-12831dc2-c6a2-4f37-99dc-ef3e005beaca.jpeg)

I’m going to lunch and in 1 hour I get back to you about feedback from the new features 

thanks :)

**Reply by FelipeRearden** [reply]
@publicus 

The new amazing feature is running perfect in 1.2.0 on Obsidian 0.12.10 on MacOS

For me, this issue is **closed**

Thanks !!!
